text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+
+### Gossiping
+
+The gossiper undergoes a few changes starting with no more generics. Existing functionality can be kept and the component can conditionally announce objects that are incoming deploys still.
+
+The gossiper supports an effect in the form of
+
+```rust
+/// Gossip a set of objects.
+///
+/// Dependencies of objects will not be retrieved, checked or gossiped,
+/// it is up to the caller to ensure the object itself and its dependencies
+/// are in the local storage.
+pub async fn gossip_objects(self, objects: Vec<Object>);
+```
+
+The algorithm works otherwise similar to how it works now.
+
+### Fetching
+
+The fetcher component also loses its type parameter, similar to the gossiper.
+
+```rust
+/// Retrieve an object from storage or the network.
+///
+/// Retrieves an object
+pub async fn fetch_object(self, object_hash: Object,
+                                max_depth: Option<NonZeroUsize>,
+                                peer: I) -> HashMap<ObjectHash, <FetchResult>>;
+```
+
+Here we employ the same `max_depth` limiter as we do with the storage itself.
+
+It is worth noting that given a current block hash `h`, calling `fetch_object(h, None, some_peer)` will eventually fetch the entire state required to join the network!
+
+### DAG/Consensus
+
+There are two ways of embedding consensus-related DAG data into the system, either the object format accompanies every single type of supported consensus with an enum variant, or a generic variant is used.
+
+The first alternative is not as a bad as it sounds; while it does constitute a leaky abstraction, it improves the general robustness of the system, which will know how to reject invalid node types and can handle them "properly" inside its storage.
+
+The second alternative is adding a generic
+
+```rust
+enum Object {
+    // ...
+    DagNode<Box<dyn DagNode>>
+}
+```
+
+variant and implementing variations of these inside the consensus component. This will result in a trait and the consensus level having to do a bit of serialization/deserialization, but does not leak the associated discriminator of the enum into the serialized representation.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+The biggest issue is that changes might affect execution engine performance negatively. This is still open to discovery, see [unresolved questions](#unresolved-questions).
+
+There is a considerable effort required to update all components and the change might not be easy to implement incrementally.
+
+The whole change to the execution engine may be prohibitive because of external dependencies, i.e. consumers of said execution engine that are not our node. This _might_ be worked around with a trait that abstracts the coupling between `Object` and components away.
```

Comment:
> The whole change to the execution engine may be prohibitive because of external dependencies, i.e. consumers of said execution engine that are not our node. This _might_ be worked around with a trait that abstracts the coupling between `Object` and components away.

It isn't apparent to me that this proposal actually affects global state at all; as far as I can tell the TrieStore would remain untouched. Please clarify?

Reply:
Ideally `TrieStore` store elements would also find their place in the unified object store, but this needs to be examined still =(